src/trace-writer.ts

```diff
@@ -121,102 +121,87 @@ export class TraceWriter extends common.Service {
     this.isActive = false;
   }
 
-  initialize(cb: (err?: Error) => void): void {
-    // Ensure that cb is called only once.
-    let pendingOperations = 2;
+  getConfig(): TraceWriterConfig {
+    return this.config;
+  }
 
+  async initialize(): Promise<void> {
     // Schedule periodic flushing of the buffer, but only if we are able to get
     // the project number (potentially from the network.)
-    this.getProjectId().then(
-        () => {
-          this.scheduleFlush();
-          if (--pendingOperations === 0) {
-            cb();
-          }
-        },
-        (err: Error) => {
-          this.logger.error(
-              'TraceWriter#initialize: Unable to acquire the project number',
-              'automatically from the GCP metadata service. Please provide a',
-              'valid project ID as environmental variable GCLOUD_PROJECT, or',
-              `as config.projectId passed to start. Original error: ${err}`);
-          cb(err);
-        });
-
-    this.getHostname((hostname) => {
-      this.getInstanceId((instanceId) => {
-        // tslint:disable-next-line:no-any
-        const addDefaultLabel = (key: string, value: any) => {
-          this.defaultLabels[key] = `${value}`;
-        };
-
-        this.defaultLabels = {};
-        addDefaultLabel(
-            TraceLabels.AGENT_DATA, `node ${pjson.name} v${pjson.version}`);
-        addDefaultLabel(TraceLabels.GCE_HOSTNAME, hostname);
-        if (instanceId) {
-          addDefaultLabel(TraceLabels.GCE_INSTANCE_ID, instanceId);
-        }
-        const moduleName = this.config.serviceContext.service || hostname;
-        addDefaultLabel(TraceLabels.GAE_MODULE_NAME, moduleName);
-
-        const moduleVersion = this.config.serviceContext.version;
-        if (moduleVersion) {
-          addDefaultLabel(TraceLabels.GAE_MODULE_VERSION, moduleVersion);
-          const minorVersion = this.config.serviceContext.minorVersion;
-          if (minorVersion) {
-            let versionLabel = '';
-            if (moduleName !== 'default') {
-              versionLabel = moduleName + ':';
-            }
-            versionLabel += moduleVersion + '.' + minorVersion;
-            addDefaultLabel(TraceLabels.GAE_VERSION, versionLabel);
-          }
-        }
-        Object.freeze(this.defaultLabels);
-        if (--pendingOperations === 0) {
-          cb();
+    const getProjectIdAndScheduleFlush = async () => {
+      try {
+        await this.getProjectId();
+      } catch (err) {
+        this.logger.error(
+            'TraceWriter#initialize: Unable to acquire the project number',
+            'automatically from the GCP metadata service. Please provide a',
+            'valid project ID as environmental variable GCLOUD_PROJECT, or',
+            `as config.projectId passed to start. Original error: ${err}`);
+        throw err;
+      }
+      this.scheduleFlush();
+    };
+    const [hostname, instanceId] = await Promise.all([
```

Comment:
`const [hostname, instanceId, ignore] = ...`

Correctness should be obvious. The missing third return value seems like an oversight at first glance.

Reply:
`getProjectIdAndScheduleFlush` returns `void`, so IMO it would seem just as much an oversight to try to get the return value. Anyway, I added `_` as the third return value.